lib/rspec/matchers/built_in/compound.rb

```diff
@@ -36,6 +36,30 @@ def expects_call_stack_jump?
           NestedEvaluator.matcher_expects_call_stack_jump?(matcher_2)
         end
 
+        # @api private
+        # @return [Boolean]
+        def diffable?
+          matcher_1.diffable? || matcher_2.diffable?
+        end
+
+        # @api private
+        # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+        def expected
+          return nil unless evaluator
+
+          expecteds = ::RSpec::Matchers::ExpectedsForMultipleDiffs.new
+
+          if !matcher_1_matches? && matcher_1.diffable?
+            expecteds.add(matcher_1.expected, :for_matcher => matcher_1)
+          end
+
+          if !matcher_2_matches? && matcher_2.diffable?
+            expecteds.add(matcher_2.expected, :for_matcher => matcher_2)
+          end
```

Comment:
Personally, I prefer immutable objects when we can find a way to make that work, since it's easier to reason about.  Can we make `RSpec::Matchers::ExpectedsForMultipleDiffs` immutable?